lib/logical_solver/presolve.cpp

```diff
@@ -0,0 +1,889 @@
+      feasible_value = getVariableFeasibleValueMIP(
+        row_index, col_index, problem_matrix_.at(row_index).at(col_index), 
+        feasibleValueCalculationBound
+      );
+    } else {
+      feasible_value = feasibleValueCalculationBound/variable_coeff;
+    }
+
+    if (feasible_value == kInfinity) {infeasible = true;}
+    else {
+      feasible_solution.at(col_index) = feasible_value;
+      postsolve_active_cols_.at(col_index) = true;
+      // Check if we have found feasible values for all variables in the 
+      // row. If so, check if the row (constraint) is satisfied, and if 
+      // so turn on row in postsolve.
+      if (isRowActivePostsolve(row_index)) {
+        if (checkConstraint(row_index, 0)) {
+          postsolve_active_rows_.at(row_index) = true;
+        }
+      }
+    }
+  };
+
+  void Presolve::updateStateRowSingletonEquality(
+    int row_index, int col_index
+  ) {
+    int variable_coefficient = problem_matrix_.at(row_index).at(col_index);
+    int RHS = lower_bounds_.at(row_index);
+    // In row singleton equality, if we are solving the MIP,
+    // we check it is feasible in presolve already so that we 
+    // do not continue the process of it is not feasible.
+    double variable_value;
+    if (solve_MIP_) {
+      variable_value = getVariableFeasibleValueMIP(
+        row_index, col_index, variable_coefficient, RHS
+      );
+    } else {
+      variable_value = (double)RHS/variable_coefficient;
+    } 
+    if (variable_value != kInfinity) {
+      presolve_active_rows_.at(row_index) = false;
+      presolve_active_rows_count_ -= 1;
+      implied_lower_bounds_.at(col_index) = variable_value;
+      implied_upper_bounds_.at(col_index) = variable_value;
+
+      // Update presolve stack.
+      struct presolve_log log = {row_index, col_index, 2};
+      presolve_stack_.push(log);
+    } else {
+      infeasible = true;
+    }
+  };
+
+  void Presolve::updateStateRowSingletonInequality(
+    int row_index, int col_index
+  ) {
+    // Get the implied bound by divinding the lower bound on the
+    // constraint by the coefficient.
+    int variable_coefficient = problem_matrix_.at(row_index).at(col_index);
+
+    if (variable_coefficient < 0) {
+      // If coefficient negative, upper bound becomes 
+      // lower bound divided by coefficient, and we use floor
+      // to make sure that the bound is met in integer arithmetic.
+      int implied_bound = std::floor(
+        float(lower_bounds_.at(row_index))/float(variable_coefficient)
+      );
+      implied_lower_bounds_.at(col_index) = -implied_upper_bounds_.at(col_index);
+      implied_upper_bounds_.at(col_index) = implied_bound;
+
+    } else {
+      // If coefficient not negative, lower bound becomes lower
+      // bound divided by the coefficient and we use ceil to make
+      // bound is met in integer arithmetic.
+      int implied_bound = std::ceil(
+        float(lower_bounds_.at(row_index))/float(variable_coefficient)
```

Comment:
Anything to avoid division by 0 errors here?

Reply:
the way the code works variable_coefficient is always non-zero, could add it at some point though.